regression_report_v2.py

```diff
@@ -0,0 +1,105 @@
+#  distributed with this work for additional information
+#  regarding copyright ownership.  The ASF licenses this file
+#  to you under the Apache License, Version 2.0 (the
+#  "License"); you may not use this file except in compliance
+#  with the License.  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+# limitations under the License.
+################################################################################
+import argparse
+import json
+import urllib
+import urllib2
+
+from regression_report import loadBenchmarkNames
+
+"""
+This is a regression detection algorithm based on the historical maximum/minimum value, please refer
+to https://docs.google.com/document/d/1Bvzvq79Ll5yxd1UtC0YzczgFbZPAgPcN3cI0MjVkIag/edit the detailed design.
+"""
+
+ENVIRONMENT = 2
+
+"""
+Returns a list of benchmark results
+"""
+def loadHistoryData(codespeedUrl, exe, benchmark, baselineSize):
+    url = codespeedUrl + 'timeline/json/?' + urllib.urlencode(
+        {'exe': exe, 'ben': benchmark, 'env': ENVIRONMENT, 'revs': baselineSize})
+    f = urllib2.urlopen(url)
+    response = f.read()
+    f.close()
+    timelines = json.loads(response)['timelines'][0]
+    result = timelines['branches']['master'][exe]
+    lessIsbBetter = (timelines['lessisbetter'] == " (less is better)")
+    return result, lessIsbBetter
+
+def detectRegression(urlToBenchmark, stds, scores, baselineSize, minRegressionRatio, minInstabilityMultiplier,
+                     direction):
+    sustainable_x = [min(scores[i - 3: i]) for i in range(3, baselineSize)]
+    baseline_throughput = max(sustainable_x)
+    current_throughput = max(scores[-3:])
+    current_instability = stds[-1] / current_throughput
+    if direction * (1 - current_throughput / baseline_throughput) > max(minRegressionRatio, minInstabilityMultiplier * current_instability):
+        print "<%s|%s> baseline=%s current_value=%s" % (urlToBenchmark, benchmark, baseline_throughput, current_throughput)
+
+def checkBenchmark(args, exe, benchmark):
+    results, lessIsbBetter = loadHistoryData(args.codespeedUrl, exe, benchmark, args.baselineSize + 3)
+    results = list(reversed(results))
+    scores = [score for (date, score, deviation, commit, branch) in results]
+    stds = [deviation for (date, score, deviation, commit, branch) in results]
+
+    urlToBenchmark = args.codespeedUrl + 'timeline/#/?' + urllib.urlencode({
+        'ben': benchmark,
+        'exe': exe,
+        'env': ENVIRONMENT,
+        'revs': args.displaySamples,
+        'equid': 'off',
+        'quarts': 'on',
+        'extr': 'on'})
+
+    if len(results) < args.baselineSize:
```

Comment:
Does this mean that a new benchmark will only be included in the regression detection 30 days after this benchmark is added?

I am wondering if it would be useful to include new benchmark in the detection faster. For example, we can start to detect regression after there are 5 samples for this benchmark.

Reply:
> Does this mean that a new benchmark will only be included in the regression detection 30 days after this benchmark is added?

Yes, a new benchmark will only be included in the regression detection 30 days after this benchmark is added.
To detect new benchmark faster, It is a good idea to change the minimum sample limit to 5.

Reply:
> Does this mean that a new benchmark will only be included in the regression detection 30 days after this benchmark is added?

Yes, a new benchmark will only be included in the regression detection 30 days after this benchmark is added.
To detect new benchmark faster, It is a good idea to change the minimum sample limit to 5.